enhance: Forward report

Johann150 opened the pull request:
# What
Add the ability for users to choose to forward a report to the other instance for remote users.

# Why
fix #8000 

# Additional info (optional)
The API is extended in a backward-compatible manner.

In the future, the API should probably be further expanded to allow to pass in URLs of notes, so the note link does not have to be in the report text.

Johann150 committed:
implement sending AP Flag object

Optionally allow a user to select to forward a report about a remote
user to the other instance. This is added in a backwards-compatible way.

Johann150 commented:
The locale strings are still missing because I don't know Japanese well enough. In English it would be:
```
forwardReport: "Forward the report to the other instance."
forwardReportIsAnonymous: "The other instance will see a system account instead of your name."
```
What should it be in Japanese?

syuilo commented:
```
forwardReport: "リモートインスタンスに通報を転送する"
forwardReportIsAnonymous: "リモートインスタンスからはあなたの情報は見れず、匿名のシステムアカウントとして表示されます。"
```

Johann150 committed:
add locale string

mei23 commented:
リモートに転送するか選択するスイッチは、ユーザーの通報画面ではなく管理画面の通報一覧にあるべきと思ってる。

Johann150 commented:
I put it on the users screen because that is how [Pleroma](https://api.pleroma.social/#operation/ReportController.create) and [Mastodon](https://docs.joinmastodon.org/methods/accounts/reports/) do it.

mei23 commented:
I know that. I think Mastodon is strange.

Johann150 commented:
Trying to find out why it is implemented this way in Pleroma and Mastodon, I found that Pleroma copied it from Mastodon. And Mastodon has this pull request: https://github.com/mastodon/mastodon/pull/6570

Looking further I found this comment that actually suggested that only the admin can forward a report: <https://github.com/mastodon/mastodon/issues/2176#issuecomment-310358890>. So I agree it would probably be a good idea to instead add the ability to the reports panel so the admin can select it. It should be selectable for each report separately.

I am not sure how to best implement this though. I could imagine
1. A button that immediately forwards the report. We would have to save whether the report has been forwarded already so that it can not be forwarded multiple times.
2. A button that forwards and closes the report.
3. A switch that forwards the report when the moderator clicks on "mark the issue as resolved". Then we would not have to save whether the report was forwarded already.

Johann150 committed:
forward report only for moderators

Johann150 committed:
add switch to moderator UI to forward report

Johann150 committed:
fix report note url

Johann150 commented:
I chose option 3, but still saving whether the report was forwarded so that this can be seen for resolved reports.

A moderator can change the switch if they want to forward the report (off by default), and then click on the mark as resolved button.
<details>
<summary>screenshot of a new report</summary>

![image](https://user-images.githubusercontent.com/20990607/144462902-5cd9c2f6-9083-4554-a9cd-6ad22529fe29.png)
</details>

An admin on the other instance will see something like the next screenshot. Note that the toggle switch is off and disabled because the report can not be forwarded again.
<details>
<summary>screenshot of a forwarded report on the other instance</summary>

![image](https://user-images.githubusercontent.com/20990607/144462146-cb4a9686-c066-4146-af41-24b949c2a5bf.png)
</details>

I changed the UI of resolved reports slightly. I use the existing `$ts.moderator` i18n string for asignee and added a `$ts.reporter` i18n string. I moved the assignee information to the same position as the reporter information to make it easier to see at a glance.

The switch to forward a report will still be visible for resolved reports, but of course disabled because it can not be changed after resolving the report. It will be on or off to show whether the report was forwarded. In this screenshot it was forwarded so it is on.
<details>
<summary>screenshot of a forwarded report on the instance the report was created on</summary>

![image](https://user-images.githubusercontent.com/20990607/144463258-674f1a01-cd91-421c-9c94-86d4de271d8d.png)
</details>

For local users, forwarding the report will of course not be available. It will look like this (this time in Japanese):

<details>
<summary>screenshot of a report concerning a local user</summary>

![image](https://user-images.githubusercontent.com/20990607/144478466-da4634cc-1780-4cff-a184-ab62ef1b0baf.png)
</details>

Johann150 commented:
Rebased to resolve the merge conflict. I don't think the test failure is caused by this code, the test passed earlier.

Johann150 committed:
return forwarded status from API

apparently forgot to carry this over from my testing environment

syuilo reviewed (approved):
YOSASOU

Johann150 commented:
Installed this on my own instance yesterday and tested with someone else that it is working with Mastodon. 🎉

Although they noted that the report looks like it is regarding the whole account and not a specific note, which may cause issues with the way moderation tools work in Mastodon: Moderators would only be able to do things with the account and not with the specific post. I am pretty sure this is due to the value of `object` in the `Flag` activity that is sent (2nd argument to `renderFlag`). This should also include the URL of the specific note, but that would require changing of `@/components/abuse-report-window.vue` to pass the URL of the note along.

But since reporting works in general I think this is something that can be fixed later. We can create a separate issue when this is merged.

Johann150 commented on packages/backend/migration/1637320813000-forwarded-report.js:
```diff
@@ -0,0 +1,13 @@
+const { QueryRunner } = require('typeorm');
+
+module.exports = class forwardedReport1637320813000 {
+	name = 'forwardedReport1637320813000';
```
Will this migration apply correctly if there is another migration with a higher number? `1639325650583-chart-v3.js` was merged yesterday

Johann150 commented:
It seems this is currently broken 😞

Johann150 committed:
object in Flag activity has to be an array

For correct interoperability with Pleroma the "object" property of the Flag
activity has to be an array.

This array will in the future also hold the link to respective notes, so it
makes sense to correct this on our side.

Johann150 commented:
With the latest commit, it will also work correctly with Pleroma. 🥳

syuilo committed:
Update get-note-menu.ts

syuilo commented:
🆒
